libraries/chain/db_maint.cpp

```diff
@@ -922,12 +922,42 @@ void database::perform_chain_maintenance(const signed_block& next_block, const g
       d.accounts_registered_this_interval = 0;
    });
 
-   // Reset all BitAsset force settlement volumes to zero
-   for( const auto& d : get_index_type<asset_bitasset_data_index>().indices() )
-   {
-      modify( d, [](asset_bitasset_data_object& o) { o.force_settled_volume = 0; });
-      if( d.has_settlement() )
-         process_bids(d);
+   if(head_block_time() >= HARDFORK_CORE_518_TIME) {
+      for( const auto& d : get_index_type<asset_bitasset_data_index>().indices() ) {
+         modify(d, [this](asset_bitasset_data_object &o) {
+            o.force_settled_volume = 0; // Reset all BitAsset force settlement volumes to zero
+            // Check if asset is smartcoin
+            if (o.feeds.size() > 0) {
+               const auto &settlement_price = o.feeds.begin()->second.second.settlement_price;
+               if (!settlement_price.is_null()) {
+                  const auto &asset = get(settlement_price.base.asset_id);
+                  auto flags = asset.options.flags;
+                  if ((flags & witness_fed_asset) || (flags & committee_fed_asset)) { // if smartcoin
+                     // check overflow
+                     if (std::numeric_limits<uint32_t>::max() - o.options.feed_lifetime_sec >
+                         head_block_time().sec_since_epoch()) {
+                        fc::time_point_sec calculate = head_block_time() - o.options.feed_lifetime_sec;
+                        for (auto itr = o.feeds.rbegin(); itr != o.feeds.rend();) { // loop feeds
+                           auto feed_time = itr->second.first;
+                           std::advance(itr, 1);
+                           if (feed_time < calculate)
+                              o.feeds.erase(itr.base()); // delete expired feed
+                        }
+                     }
+                  }
+               }
+            }
+         });
+         if (d.has_settlement())
+            process_bids(d);
```

Comment:
Hmm, this is a bit ugly (duplicate code). But I guess it's better to live with it at this moment.

Reply:
@pmconrad it's a side-effect of one of my earlier requests, since hard fork time check is independent, for slightly better performance, moved it to out of `for`.